### PR TITLE
Add a public "has" method to the Controller class, to check the existence of a given controller

### PR DESCRIPTION
**Pull request by Vadorequest**
I haven't written any test. The `npm install` failed here and I've been unable to test anything. (Seems like it fails on gyp-rebuild with my WIndows 8.1...)

The PR is quite small. if someone with a proper setup could just add some test for it, it would be nice.

https://github.com/angular/angular.js/issues/13951

### CODE REVIEW

File: src/ng/controller.js [hunk 1]
```diff
@@ -29,6 +29,15 @@ function $ControllerProvider() {
 
   /**
    * @ngdoc method
+   * @name $controllerProvider#has
+   * @param (string) name Controller name to check.
```
Comment (gkalpak) [inline src/ng/controller.js]: Use `{...}` instead of `(...)`

File: src/ng/controller.js [hunk 2]
```diff
@@ -29,6 +29,15 @@ function $ControllerProvider() {
...
+   */
+  this.has = function(name){
+    return (name && controllers[name])? true : false;
```
Comment (gkalpak) [inline src/ng/controller.js]: Use `controllers.hasOwnProperty(name)` to avoid looking up the prototype chain.

**Comment by gkalpak** [comment]
I left a couple of minor comments. There are also some styling issues that JSCS complains about (you can take a look [here][1]).

Could you also add `Fixes #13951` at the end of your commit message ? This way, the related issue will be automatically closed when this PR is merged.

--
Regarding `npm install`:
Indeed `gyp-rebuild` will throw lots of read in your face, while trying to rebuild some modules on Windows. This doesn't necessarily mean the installation failed (it will just fall back to pre-built artifacts).
Are you sure `npm install` failed ? (Even if it has, the `gyp-rebuild` failures might not be the cause.)

Regarding tests:
Even if you are not able to run the tests locally, you could still add them and push the changes to this PR, so CI can run them for you. It's not necessary though (and it's tooo slow and obviously less than ideal :smiley:).

[1]: https://travis-ci.org/angular/angular.js/jobs/111074010

**Comment by Vadorequest** [comment]
@gkalpak Thanks for feedback. I've pushed changes. I'm installing the environment on an Ubunutu now. I don't know what happened about gyp-rebuild but many node modules weren't properly installed after the red messages, I assumed it didn't go well. :)

(I forgot about `Fixes #13951` but since I'll eventually have to `rebase -i` I'll add it later anyway.

**Comment by gkalpak** [comment]
Great ! The code LGTM. Let us know if you run into any difficulties setting up the testing environment.

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperties

**Comment by Vadorequest** [comment]
I just pushed some tests, hope they'll pass. Not sure aboute the `it` messages though.
I definitely ran into issues with the setup, I got red messages again after a `npm install`, but all packages seemed to be installed, then the whole `node_modules` folder got deleted. I really don't get it. Seems like an exception has been thrown somewhere and the whole thing got cancelled, I tried twice, same result.

So, CI's got to do the testing part :)

**Comment by gkalpak** [comment]
Are you by any chance using npm >=3.0.0 ? That would explain the `node_modules/` directory getting deleted after modules are installed. It's because the execution order of pre/post-install scripts wrt to other operations has changed in npm 3.
Our setup is currently not compatible with npm 3 (and even if you worked around that, I doubt all our dependencies are). This is tracked in #13505 btw.

**Comment by gkalpak** [comment]
There were some [failures][1] :smiley:
Taking a closer look...

[1]: https://travis-ci.org/angular/angular.js/builds/111180072

File: test/ng/controllerSpec.js [hunk 3]
```diff
@@ -57,6 +57,27 @@ describe('$controller', function() {
...
 
+    it('should return true when having an existing controller', function() {
+      var FooCtrl = function($scope) { $scope.foo = 'foo'; },
+          scope = {},
+          ctrl;
+
+      $controllerProvider.register({FooCtrl: FooCtrl});
+
+      ctrl = $controller('FooCtrl', {$scope: scope});
+      expect(ctrl.has('FooCtrl')).toBe(true);
```
Comment (gkalpak) [inline test/ng/controllerSpec.js]: It's `$controllerProvider` that has the `.has()` method. Not specific controller instances.
This test could look like:

```js
$controllerProvider.register('FooCtrl', noop);

expect($controllerProvider.has('FooCtrl')).toBe(true);
expect($controllerProvider.has('BarCtrl')).toBe(false);
```

Or if we wanted to be more thorough, something like:

```js
$controllerProvider.register('FooCtrl', noop);
$controllerProvider.register('BarCtrl', ['dep1', 'dep2', noop]);
$controllerProvider.register({
  'BazCtrl': noop,
  'QuxCtrl: ['dep1', 'dep2', noop]
});

expect($controllerProvider.has('FooCtrl')).toBe(true);
expect($controllerProvider.has('BarCtrl')).toBe(true);
expect($controllerProvider.has('BazCtrl')).toBe(true);
expect($controllerProvider.has('QuxCtrl')).toBe(true);

expect($controllerProvider.has('UnknownCtrl')).toBe(false);
```

And basically, the next test would be redundant.

File: test/ng/controllerSpec.js [hunk 4]
```diff
@@ -57,6 +57,19 @@ describe('$controller', function() {
...
+      $controllerProvider.register({FooCtrl: FooCtrl});
+      expect($controllerProvider.has('FooCtrl')).toBe(true);
+    });
+
+    it('should return false when not having a non existing controller', function() {
+      var FooCtrl = function($scope) { $scope.foo = 'foo'; };
+
+      $controllerProvider.register({FooCtrl: FooCtrl});
+      expect($controllerProvider.has('BarCtrl')).toBe(false);
+    });
```
Comment (gkalpak) [inline test/ng/controllerSpec.js]: I would basically merge both tests, into something like:

```js
$controllerProvider.register('FooCtrl', noop);

expect($controllerProvider.has('FooCtrl')).toBe(true);
expect($controllerProvider.has('BarCtrl')).toBe(false);
```

Or if we wanted to be more thorough, something like:

```js
$controllerProvider.register('FooCtrl', noop);
$controllerProvider.register('BarCtrl', ['dep1', 'dep2', noop]);
$controllerProvider.register({
  'BazCtrl': noop,
  'QuxCtrl': ['dep1', 'dep2', noop]
});

expect($controllerProvider.has('FooCtrl')).toBe(true);
expect($controllerProvider.has('BarCtrl')).toBe(true);
expect($controllerProvider.has('BazCtrl')).toBe(true);
expect($controllerProvider.has('QuxCtrl')).toBe(true);

expect($controllerProvider.has('UnknownCtrl')).toBe(false);
```

**Comment by gkalpak** [comment]
@Vadorequest, do you think it would be useful to have the method available on `$controller` as well ?

**Comment by Vadorequest** [comment]
@gkalpak No, I don't think the controller needs to have a `has` method, the provider makes much more sense.

File: test/ng/controllerSpec.js (hunk 4 above)
Comment (Vadorequest) [inline test/ng/controllerSpec.js]: I don't understand what `noop` is here, should I replace it by `{}`?

Comment (gkalpak) [inline test/ng/controllerSpec.js]: Sorry, too much internals :grin:
`noop` is basically `angular.noop`, which is basically `function() {}` (a no-op function). You can use `noop` as a shorthand for an empty constructor function. For this functionality, you don't need to pass any arguments to the controller, nor does the controller need to have any logic.

**Comment by Vadorequest** [comment]
Tests passed :) Cool.

And yes, I'm using node 5.5, that's why it fails during the setup.

**Comment by Vadorequest** [comment]
Using node 3.3.1 it worked well, I've been able to run the tests and everything looks good.
The PR has been squashed and is ready to be merged.

File: test/ng/controllerSpec.js [hunk 5]
```diff
@@ -57,6 +57,24 @@ describe('$controller', function() {
       expect(ctrl instanceof BarCtrl).toBe(true);
     });
 
+    it('should return true when having an existing controller, should return false otherwise', function() {
+      var FooCtrl = function($scope) { $scope.foo = 'foo'; },
+          BarCtrl = function($scope) { $scope.bar = 'bar'; };
```
Comment (gkalpak) [inline test/ng/controllerSpec.js]: We don't need these two controllers, do we ?

**Comment by gkalpak** [comment]
@Vadorequest, only a minor clean-up comment. LGTM otherwise :+1:

File: test/ng/controllerSpec.js (hunk 5 above)
Comment (Vadorequest) [inline test/ng/controllerSpec.js]: Indeed! I just removed them.

**Comment by gkalpak** [comment]
Th @Vadorequest :+1: 
Backported to `v1.5.x` as bb9575d.

**Comment by Vadorequest** [comment]
Thanks to you @gkalpak :)
Does that mean that it can be used in the latest version righ now?

**Comment by gkalpak** [comment]
It will be available from the upcoming `v1.5.1` release onwards. Right now you can try it out with the [latest snapshot][1], which is basically the code urrently on master.
(NOTE: The snapshot changes with every commit to master and might contain breaking changes. It is **not** recommened for use in production :smiley: - it's just an easy way to try something out before it's released.)

[1]: https://code.angularjs.org/snapshot/

**Comment by Vadorequest** [comment]
Thanks ;)

**Pull request closed without merging**